quill-jdbc-zio/src/main/scala/io/getquill/context/zio/ZioJdbcContext.scala

```diff
@@ -0,0 +1,231 @@
+  override type RunBatchActionReturningResult[T] = List[T]
+
+  // Need explicit return-type annotations due to scala/bug#8356. Otherwise macro system will not understand Result[Long]=Task[Long] etc...
+  override def executeAction[T](sql: String, prepare: Prepare = identityPrepare): RIO[Connection, Long] =
+    super.executeAction(sql, prepare)
+  override def executeQuery[T](sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): RIO[Connection, List[T]] =
+    super.executeQuery(sql, prepare, extractor)
+  override def executeQuerySingle[T](sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): RIO[Connection, T] =
+    super.executeQuerySingle(sql, prepare, extractor)
+  override def executeActionReturning[O](sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[O], returningBehavior: ReturnAction): RIO[Connection, O] =
+    super.executeActionReturning(sql, prepare, extractor, returningBehavior)
+  override def executeBatchAction(groups: List[BatchGroup]): RIO[Connection, List[Long]] =
+    super.executeBatchAction(groups)
+  override def executeBatchActionReturning[T](groups: List[BatchGroupReturning], extractor: Extractor[T]): RIO[Connection, List[T]] =
+    super.executeBatchActionReturning(groups, extractor)
+  override def prepareQuery[T](sql: String, prepare: Prepare, extractor: Extractor[T] = identityExtractor): Connection => RIO[Connection, PreparedStatement] =
+    super.prepareQuery(sql, prepare, extractor)
+  override def prepareAction(sql: String, prepare: Prepare): Connection => RIO[Connection, PreparedStatement] =
+    super.prepareAction(sql, prepare)
+  override def prepareBatchAction(groups: List[BatchGroup]): Connection => RIO[Connection, List[PreparedStatement]] =
+    super.prepareBatchAction(groups)
+
+  override protected val effect: Runner = Runner.default
+  import effect._
+
+  /** ZIO Contexts do not managed DB connections so this is a no-op */
+  override def close(): Unit = ()
+
+  protected def withConnection[T](f: Connection => Result[T]): Result[T] = throw new IllegalArgumentException("Not Used")
+  override protected def withConnectionWrapped[T](f: Connection => T): RIO[Connection, T] =
+    RIO.fromFunction(f) // TODO fromFunctionM and ZIO.effectBlocking?
+
+  trait SameThreadExecutionContext extends scala.concurrent.ExecutionContext {
+    def submit(runnable: Runnable): Unit =
+      runnable.run()
+  }
+
+  def transaction[A](f: RIO[Connection, A]): RIO[Connection, A] = {
+    def die = Task.die(new IllegalStateException("The task was cancelled in the middle of a transaction."))
+    ZIO.environment[Connection].flatMap(conn =>
+      f.onInterrupt(die).onExit {
+        case Success(_) =>
+          UIO(conn.commit())
+        case Failure(cause) =>
+          // TODO Are we really catching the result of the conn.rollback() Task's exception?
+          catchAll(Task(conn.rollback()) *> Task.halt(cause))
```

Comment:
Instead of catching rollback errors, maybe one possible alternative could be to combine the causes (like with `++` for example) since `Cause` can represent multiple failures.